### PR TITLE
Integrate GET user-resources (VMs) api into Guacamole

### PR DESCRIPTION
**Pull request by anatbal**
# PR for issue #721 , #751

## What is being addressed

- Add a call to GET user-resources in order to get the vm list.
-  Fix a bug when fetching the username and password (remove whitespace)
- Add service ID to the terraform and webapp

### CODE REVIEW

**Review by msebragge — APPROVED** [review]

**Review by tamirkamara — CHANGES_REQUESTED** [review]
Placing a hold here

File: templates/workspace_services/guacamole/porter.yaml [hunk 1]
```diff
@@ -126,6 +126,7 @@ install:
         guac_disable_download: "{{ bundle.parameters.guac_disable_download }}"
         api_client_id: "{{ bundle.parameters.api_client_id }}"
         aad_tenant_id: "{{ bundle.parameters.aad_tenant_id }}"
+        key: "{{ bundle.parameters.id }}"
```
Comment (tamirkamara) [inline templates/workspace_services/guacamole/porter.yaml:129]: I think the name "key" can be renamed to something more meaningful for terraform in this case. What about something like tre_resource_id?

File: templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java [hunk 2]
```diff
@@ -88,64 +103,43 @@
...
-        String json = EntityUtils.toString(httpResponse.getEntity());
-        if (json.length() != 0) {
-          virtualMachines = new JSONArray(json);
-        } else {
-          virtualMachines = new JSONArray();
+        try {
+            final CloseableHttpClient httpClient = HttpClients.custom()
+                .setSSLSocketFactory(new SSLConnectionSocketFactory(
+                    SSLContexts.custom().loadTrustMaterial(null, new TrustSelfSignedStrategy()).build(),
+                    NoopHostnameVerifier.INSTANCE)
```
Comment (tamirkamara) [inline templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java:112]: This is something we should figure out...

File: templates/workspace_services/guacamole/terraform/locals.tf [hunk 3]
```diff
@@ -7,10 +7,12 @@ resource "random_string" "unique_id" {
...
   service_id                   = random_string.unique_id.result
   short_workspace_id           = substr(var.workspace_id, -4, -1)
+  location                     = data.azurerm_resource_group.ws.location
   service_resource_name_suffix = "${var.tre_id}-ws-${local.short_workspace_id}-svc-${local.service_id}"
   webapp_name                  = "guacamole-${local.service_resource_name_suffix}"
   core_vnet                    = "vnet-${var.tre_id}"
   core_resource_group_name     = "rg-${var.tre_id}"
   issuer                       = "https://login.microsoftonline.com/${var.aad_tenant_id}/v2.0"
   kv_url                       = "https://kv-guac-${var.tre_id}-${local.short_workspace_id}.vault.azure.net"
+  project_url                  = "https://${var.tre_id}.${local.location}.cloudapp.azure.com"
```
Comment (tamirkamara) [inline templates/workspace_services/guacamole/terraform/locals.tf:17]: Still open and need a decision if we access the api through the app-gw or directly.

**Review by tamirkamara — CHANGES_REQUESTED** [review]

File: templates/workspace_services/guacamole/porter.yaml (hunk 1 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/porter.yaml]: Fixed

File: templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java (hunk 2 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java]: Decided on the "internal" approach. Changed accordingly

File: templates/workspace_services/guacamole/terraform/locals.tf (hunk 3 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/locals.tf]: Decided on the "internal" approach. Changed accordingly

File: templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java [hunk 4]
```diff
@@ -88,64 +102,42 @@
...
-
-        String json = EntityUtils.toString(httpResponse.getEntity());
-        if (json.length() != 0) {
-          virtualMachines = new JSONArray(json);
-        } else {
-          virtualMachines = new JSONArray();
+        try {
+            SSLContextBuilder builder = new SSLContextBuilder();
+            builder.loadTrustMaterial(null, new TrustSelfSignedStrategy());
+            SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(builder.build());
```
Comment (tamirkamara) [inline templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java:110]: Is it required to adjust the SSL things to make a secure HTTPS call?

File: templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java [hunk 5]
```diff
@@ -88,64 +102,42 @@
...
-        httpget.addHeader("Authorization", "Bearer " + user.getAccessToken());
-
-        CloseableHttpResponse httpResponse = httpclient.execute(httpget);
-
-        String json = EntityUtils.toString(httpResponse.getEntity());
-        if (json.length() != 0) {
-          virtualMachines = new JSONArray(json);
-        } else {
-          virtualMachines = new JSONArray();
+        try {
```
Comment (tamirkamara) [inline templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java:107]: Overall, seems a bit complicated than the "native" java 11 method of using HttpClient. Something like this (not 100% complete):
```java
        String url = String.format("%s/api/workspace-services/%s/user-resources", System.getenv("API_URL"), System.getenv("SERVICE_ID"));

        // create a client
        var client = HttpClient.newHttpClient();

        // create a request
        var request = HttpRequest.newBuilder(URI.create(url))
                .header("accept", "application/json")
                .header("Authorization", "Bearer " + user.getAccessToken())
                .timeout(Duration.ofSeconds(5))
                .build();

        // use the client to send the request
        var response = client.send(request, HttpResponse.BodyHandlers.ofString());

        // the response:
        if (!response.body().isBlank()) {
            final JSONObject result = new JSONObject(response.body());
            virtualMachines = result.getJSONArray("userResources");
        } else {
            virtualMachines = new JSONArray();
        }
```

**Review by tamirkamara — CHANGES_REQUESTED** [review]

File: templates/workspace_services/guacamole/terraform/locals.tf [hunk 6]
```diff
@@ -7,10 +7,12 @@ resource "random_string" "unique_id" {
 locals {
   service_id                   = random_string.unique_id.result
   short_workspace_id           = substr(var.workspace_id, -4, -1)
+  location                     = data.azurerm_resource_group.ws.location
```
[the remark this answers is not in the document]
Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/locals.tf]: No, I removed it.

File: templates/workspace_services/guacamole/terraform/web_app.tf [hunk 7]
```diff
@@ -32,6 +32,8 @@ resource "azurerm_app_service" "guacamole" {
 
     TENANT_ID    = data.azurerm_client_config.current.tenant_id
     KEYVAULT_URL = "${local.kv_url}"
+    API_URL      = "${local.api_url}"
+    SERVICE_ID   = "${var.tre_resource_id}"
```
[the remark this answers is not in the document]
Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/web_app.tf]: Fixed.

File: templates/workspace_services/guacamole/terraform/web_app.tf [hunk 8]
```diff
@@ -187,3 +189,31 @@ resource "azurerm_private_endpoint" "guacamole" {
...
   }
 }
+
+resource "azurerm_network_security_rule" "allow-outbound-within-guacamole-and-api-subnets" {
+  access                      = "Allow"
+  destination_port_range      = "443"
+  destination_address_prefix  = data.azurerm_subnet.web_apps_core.address_prefix
+  source_address_prefix       = data.azurerm_subnet.web_apps.address_prefix
+  direction                   = "Outbound"
+  name                        = "outbound-within-guacamole-and-api-subnets"
```
[the remark this answers is not in the document]
Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/web_app.tf]: Fixed.

File: templates/workspace_services/guacamole/terraform/web_app.tf [hunk 9]
```diff
@@ -187,3 +189,31 @@ resource "azurerm_private_endpoint" "guacamole" {
...
+  source_port_range           = "*"
+}
+
+resource "azurerm_network_security_rule" "allow-inbound-within-webapps-and-services-subnets" {
+  access                      = "Allow"
+  destination_port_range      = "3389"
+  destination_address_prefix  = data.azurerm_subnet.services.address_prefix
+  source_address_prefix       = data.azurerm_subnet.web_apps.address_prefix
+  direction                   = "Inbound"
+  name                        = "inbound-within-webapps-and-services-subnets"
```
[the remark this answers is not in the document]
Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/web_app.tf]: Fixed.

File: templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java (hunk 4 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java]: Following the next comment this was deleted.

File: templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java (hunk 5 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java:113]: Fixed. Actually worked pretty much out of the box.

Comment (anatbal) [reply templates/workspace_services/guacamole/guacamole-server/guacamole-auth-azure/src/main/java/org/apache/guacamole/auth/azuretre/connection/ConnectionService.java:113]: I fixed all the other comments and tested e2e (created new bundles and guacamole image, created new ws, new guacamole, new vm, saw it connecting)

**Review by tamirkamara — APPROVED** [review]
LGTM

**Pull request merged**